torch/distributions/wishart.py

```diff
@@ -0,0 +1,233 @@
+        :attr:`scale_tril` can be specified.
+        Using :attr:`scale_tril` will be more efficient: all computations internally
+        are based on :attr:`scale_tril`. If :attr:`covariance_matrix` or
+        :attr:`precision_matrix` is passed instead, it is only used to compute
+        the corresponding lower triangular matrices using a Cholesky decomposition.
+        'torch.distributions.LKJCholesky' is a restricted Wishart distribution.[1]
+
+    **References**
+
+    [1] `On equivalence of the LKJ distribution and the restricted Wishart distribution`,
+    Zhenxun Wang, Yunan Wu, Haitao Chu.
+    """
+    arg_constraints = {
+        'covariance_matrix': constraints.positive_definite,
+        'precision_matrix': constraints.positive_definite,
+        'scale_tril': constraints.lower_cholesky,
+        'df': constraints.greater_than(0),
+    }
+    support = constraints.positive_definite
+    has_rsample = True
+
+    def __init__(self,
+                 df: torch.Tensor,
+                 covariance_matrix: torch.Tensor = None,
+                 precision_matrix: torch.Tensor = None,
+                 scale_tril: torch.Tensor = None,
+                 validate_args=None):
+        assert (covariance_matrix is not None) + (scale_tril is not None) + (precision_matrix is not None) == 1, \
+            "Exactly one of covariance_matrix or precision_matrix or scale_tril may be specified."
+
+        if scale_tril is not None:
+            assert scale_tril.dim() > 1, \
+                "scale_tril must be at least two-dimensional, with optional leading batch dimensions"
+            batch_shape = torch.broadcast_shapes(scale_tril.shape[:-2], df.shape)
+            event_shape = scale_tril.shape[-2:]
+            self.scale_tril = scale_tril.expand(batch_shape + (-1, -1))
+        elif covariance_matrix is not None:
+            assert covariance_matrix.dim() > 1, \
+                "covariance_matrix must be at least two-dimensional, with optional leading batch dimensions"
+            batch_shape = torch.broadcast_shapes(covariance_matrix.shape[:-2], df.shape)
+            event_shape = covariance_matrix.shape[-2:]
+            self.covariance_matrix = covariance_matrix.expand(batch_shape + (-1, -1))
+        elif precision_matrix is not None:
+            assert precision_matrix.dim() > 1, \
+                "precision_matrix must be at least two-dimensional, with optional leading batch dimensions"
+            batch_shape = torch.broadcast_shapes(precision_matrix.shape[:-2], df.shape)
+            event_shape = precision_matrix.shape[-2:]
+            self.precision_matrix = precision_matrix.expand(batch_shape + (-1, -1))
+
+        self.df = df.expand(batch_shape)
+        self.arg_constraints['df'] = constraints.greater_than(event_shape[-1] - 1)
+        if self.df.lt(event_shape[-1]).any():
+            warnings.warn("Low df values detected. Singular samples are highly likely occured for df < ndim.")
```

Comment:
nit: occured -> to occur.

Does this commonly happen?

Reply:
Thanks.
I am not sure since it depends on not only the df values but also parameter shapes (or dimensions of square matrix).
But after adding the warning, I shows warining always when I sample 10000+ samples for 10+ dimensions and df=ndim.
I am checking the code of `scipy.stats.wishart` to mitigate the problem.

Reply:
I applied resampling procedure for now. I will look for more fundamental solution.